packages/firestore/src/core/component_provider.ts

```diff
@@ -21,8 +21,8 @@ import {
   SharedClientState,
   WebStorageSharedClientState
 } from '../local/shared_client_state';
-import { LocalStore } from '../local/local_store';
-import { SyncEngine } from './sync_engine';
+import { LocalStore, MultiTabLocalStore } from '../local/local_store';
+import { MultiTabSyncEngine, SyncEngine } from './sync_engine';
```

Comment:
This review doesn't include `MultiTabSyncEngine` but I have some thoughts on how to move `SharedClientState` and related properties into it. Most of these options will make the code ugly, awkward or might not be viable but want to throw it out there anyway.

In any method in `SyncEngine` that has multi-tab logic is it possible to refactor it to **not** be in the middle? For example `removeAndCleanupTarget()` only has one line at the top. Could you do that to everything else? If so, you could rely on method overloading in `MultiTabSyncEngine` to do pre/post processing and call the base class method in the appropriate place. If the logic can't be split it might be worth overloading the method entirely and relying on unit tests to ensure consistency.

Alternatively, in methods where you need a return value from `SharedClientState` you can modify the method to have an optional callback parameter that usually gives you a default value but in `MultiTabSyncEngine` you can overload the method and call into the base with an argument for the callback that gives you a real result. You can do the same for notification only methods OR raise an event that `MultiTabSyncEngine` can subscribe to. The latter is super weird but eliminates the reference entirely without needing a stub implementation.

Reply:
I looked through all the calls to SharedClientState to see if we can improve the logic and remove some of the tight coupling. Unfortunately, I think in almost all cases the current logic is simpler than what we would have if we pulled out the code:

- If we overwrite the methods to extract one line, it creates an interaction model that is much harder to follow. While it is relatively straightforward to pull out the first or last line in some of these methods, it essentially hides these lines from view if the use is not aware that they are overwritten in `MultiTabSyncEngine`. 
- In some cases (like `removeAndCleanupTarget()`) this is not actually all that bad - but it creates some inconsistencies between target allocation and target cleanup. I would rather keep all of this code in one place.

I did create an overwrite for `applyOnlineStateChange`. This method has different behavior for Multi-Tab clients and has no "mirrored" behavior in SyncEngine.